analyzers/src/SonarAnalyzer.Common/Helpers/SonarLintXmlReader.cs

```diff
@@ -0,0 +1,93 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.IO;
+using System.Text;
+using System.Xml;
+using System.Xml.Serialization;
+using Microsoft.CodeAnalysis.Text;
+
+namespace SonarAnalyzer.Helpers;
+
+public class SonarLintXmlReader
+{
+    public static readonly SonarLintXmlReader Empty = new(null, LanguageNames.CSharp);
+
+    private readonly SonarLintXml sonarLintXml;
+    private readonly string propertyLanguage;
+
+    private bool? ignoreHeaderComments;
+    public bool IgnoreHeaderComments => ignoreHeaderComments ??= ReadBoolean(ReadProperty($"sonar.{propertyLanguage}.ignoreHeaderComments"));
+
+    private bool? analyzeGeneratedCode;
+    public bool AnalyzeGeneratedCode => analyzeGeneratedCode ??= ReadBoolean(ReadProperty($"sonar.{propertyLanguage}.analyzeGeneratedCode"));
+
+    private string[] exclusions;
+    public string[] Exclusions => exclusions ??= ReadCommaSeparatedArray(ReadProperty("sonar.exclusions"));
+
+    private string[] inclusions;
+    public string[] Inclusions => inclusions ??= ReadCommaSeparatedArray(ReadProperty("sonar.inclusions"));
+
+    private string[] globalExclusions;
+    public string[] GlobalExclusions => globalExclusions ??= ReadCommaSeparatedArray(ReadProperty("sonar.global.exclusions"));
+
+    private string[] testExclusions;
+    public string[] TestExclusions => testExclusions ??= ReadCommaSeparatedArray(ReadProperty("sonar.test.exclusions"));
+
+    private string[] testInclusions;
+    public string[] TestInclusions => testInclusions ??= ReadCommaSeparatedArray(ReadProperty("sonar.test.inclusions"));
+
+    private string[] globalTestExclusions;
+    public string[] GlobalTestExclusions => globalTestExclusions ??= ReadCommaSeparatedArray(ReadProperty("sonar.global.test.exclusions"));
```

Comment:
When this file exists, it is always used. And it is always cached. So the lazy loading is way more complicated than it needs to be - just read it all right at the constructor. And remove the backing fields.

Reply:
@pavel-mikula-sonarsource  For my understanding. You mean to delete all the private fields and do this:

```csharp
    public SonarLintXmlReader(SourceText sonarLintXml)
    {
        this.sonarLintXml = sonarLintXml == null ? SonarLintXml.Empty : ParseContent(sonarLintXml);
        Exclusions = ReadCommaSeparatedArray(ReadSettingsProperty("sonar.exclusions"));
        Inclusions = ReadCommaSeparatedArray(ReadSettingsProperty("sonar.inclusions"));
        GlobalExclusions = ReadCommaSeparatedArray(ReadSettingsProperty("sonar.global.exclusions"));
        TestExclusions = ReadCommaSeparatedArray(ReadSettingsProperty("sonar.test.exclusions"));
        TestInclusions = ReadCommaSeparatedArray(ReadSettingsProperty("sonar.test.inclusions"));
        GlobalTestExclusions = ReadCommaSeparatedArray(ReadSettingsProperty("sonar.global.test.exclusions"));
        ParametrizedRules = ReadRuleParameters();
    }
```

Reply:
to be fixed in https://github.com/SonarSource/sonar-dotnet/pull/6924

Reply:
Yes